### PR TITLE
Add greedy longest chunked palindrome decomposition

### PR DESCRIPTION
**Pull request by alexpantyukhin**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by alexpantyukhin** [comment]
It looks like there is the error in other files.

File: greedy_methods/longest_chunked_palindrome_decomposition.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    len_text = len(text)
+
+    def substrings_are_equal(
+        first_substring_index: int, second_substring_index: int, length: int
+    ) -> bool:
+        """
+        >>> text = 'aabaa'
+        >>> substrings_are_equal(0, 3, 2)
+        True
```
Comment (CaedenPH) [inline greedy_methods/longest_chunked_palindrome_decomposition.py:58]: Docstrings within functions inside functions are not tested

**Review by CaedenPH — CHANGES_REQUESTED** [review]
Looks good. Please just remove/move the doctests from the functions within functions.

Agreed, those failing tests are from other files and I will work on finding a fix.

File: greedy_methods/longest_chunked_palindrome_decomposition.py (hunk 1 above)
Comment (alexpantyukhin) [reply greedy_methods/longest_chunked_palindrome_decomposition.py:58]: it's done! thanks.

**Comment by alexpantyukhin** [comment]
@CaedenPH could you please review it? I have merged with `master` and now tests are fixed.

**Review by alleycat786 — APPROVED** [review]
nice work here💪🏻

**Comment by alexpantyukhin** [comment]
@cclauss Could you please review this PR also?

File: greedy_methods/longest_chunked_palindrome_decomposition.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    1
+    >>> longest_decomposition('m')
+    1
+    >>> longest_decomposition('aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa')
+    46
+    >>> longest_decomposition('abcabcabcabcabcabcabcabcabcabcabcabcabcabcabc')
+    15
+    """
+
+    if text is None or len(text) == 0:
```
Comment (cclauss) [inline greedy_methods/longest_chunked_palindrome_decomposition.py:47]: What is text is False or [] or {} or (,) or set()?
```suggestion
    if not text:
```

File: greedy_methods/longest_chunked_palindrome_decomposition.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+Use greedy approach. If it's possible to find the palindrome
+chunk for the current text start, then use it and apply recursion
+for the left path of string.
+
+Runtime: O(n*n)
+Space: O(1)
+"""
+
+
+def longest_decomposition(text: str) -> int:
```
Comment (cclauss) [inline greedy_methods/longest_chunked_palindrome_decomposition.py:33]: Please add a test for the empty string, None, etc.

File: greedy_methods/longest_chunked_palindrome_decomposition.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    >>> longest_decomposition('m')
+    1
+    >>> longest_decomposition('aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa')
+    46
+    >>> longest_decomposition('abcabcabcabcabcabcabcabcabcabcabcabcabcabcabc')
+    15
+    """
+
+    if text is None or len(text) == 0:
+        raise ValueError("The text should be not empty string.")
```
Comment (cclauss) [inline greedy_methods/longest_chunked_palindrome_decomposition.py:48]:
```suggestion
        raise ValueError("The text must be a non-empty string.")
```

File: greedy_methods/longest_chunked_palindrome_decomposition.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    len_text = len(text)
+
+    def substrings_are_equal(
+        first_substring_index: int, second_substring_index: int, length: int
+    ) -> bool:
+        for i in range(length):
+            if text[first_substring_index + i] != text[second_substring_index + i]:
+                return False
+
+        return True
```
Comment (cclauss) [inline greedy_methods/longest_chunked_palindrome_decomposition.py:59]:
```suggestion
         # length = min(len(first_substring_index), len(second_substring_index))
        return first_substring_index[:length] == second_substring_index[:length]
```

File: greedy_methods/longest_chunked_palindrome_decomposition.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    >>> longest_decomposition('abcabcabcabcabcabcabcabcabcabcabcabcabcabcabc')
+    15
+    """
+
+    if text is None or len(text) == 0:
+        raise ValueError("The text should be not empty string.")
+
+    len_text = len(text)
+
+    def substrings_are_equal(
```
[the remark this answers is not in the document]
Comment (cclauss) [reply greedy_methods/longest_chunked_palindrome_decomposition.py:52]: Please do not use subfunctions because they are not doctestable.

**Review by cclauss — CHANGES_REQUESTED** [review]

File: greedy_methods/longest_chunked_palindrome_decomposition.py (hunk 3 above)
Comment (alexpantyukhin) [reply greedy_methods/longest_chunked_palindrome_decomposition.py:33]: fixed

**Comment by alexpantyukhin** [comment]
@cclauss thank you!

**Pull request closed without merging**